### ISSUE TITLE
Capybara server hangs under load - react, relay, rails-api

### ISSUE THREAD

**Issue by rosskevin** [teamcapybara/capybara#1855]
## Meta
Capybara Version: 2.13.0
Driver Information: selenium-webdriver 3.3.0

## Context
We transitioned our application from rails to react/relay/rails-api.  Our cucumber/capybara tests only needed slight modification, and as we added them back into our circleci build, eventually our `selenium/standalone-chrome:3.3.0` instances would error with either `Failed to load resource: net::ERR_EMPTY_RESPONSE` (most typical) or `Net::ReadTimeout (Net::ReadTimeout)`.  When it failed, it would not be specific to a URL, it would vary based on what was being hit at a particular time.  

When we tried with the firefox image, it similarly failed at the same point with `The connection to the server was reset while the page was loading.`

We then added a `/health` check route to run direct from cucumber in an `After` step to check the server status.  It became clear that the server capybara started stopped working.

When we started a rails server `Thin` and used `Capybara.run_server = false`, then ran our tests, we succeed without a problem.

## Expected Behavior
Does not hang.

## Actual Behavior
Hangs under after consistent load.

## Steps to reproduce
I'm not sure how to reproduce this outside our app.  I can say that with our app now as react, tests run much faster and I assume server activity is at a much higher frequency.

## Alternate solutions
Configure capybara to use the `Thin` server?  Other ideas?

**Comment by rosskevin** [comment]
I've just found through SO the `Capybara.server = :puma` and via the code `:webrick`, though I see nothing in the README.  I would PR a README change, just not sure where to put it since there isn't already a section or description.

**Comment by twalpole** [comment]
The "server" thread that Capybara starts depends on what you have `Capybara.server set` to.  If you haven't set it then it defaults to Webrick which may have issues with multiple simultaneous requests.  If that is the case then you can try setting `Capybara.server = :puma` which could make a difference.   Beyond that, Capybara keeps track of the number of open requests to the "server"  (it and when `reset_sessions` is called at the end of each test it changes the current url to "about:blank" and then waits up to 60 seconds for all the currently open requests to complete.  If requests weren't completing and that was hanging I would expect it to raise a "Requests did not finish in 60 seconds" error.  Other than that, Capybara is just running your app so it could boil down to your app leaking memory/resources or something?

**Comment by twalpole** [comment]
Hmmm -- I could have sworn we had a section in there about that -- guess not (or it was deleted at some point) - You could add it as a new section after the `Drivers` section.  In Capybara 3.0 the default will be :puma since it's needed for most modern apps, but in the 2.x versions we've had to leave it as :webrick for breaking change reasons.

**Comment by rosskevin** [comment]
Thanks @twalpole - I researched leaks first thing (I'm two weeks into this) and only yesterday did I actually figure out there was a hanging issue.  In the past few minutes, actually since switching to puma it complained about a bad `conversion of Symbol to String`, I tracked that down to rack-attack!  Since switching to circleci, we moved our browser to docker and it is no longer on the same IP address, so the rails server was throttling since it wasn't on the whitelisted localhost.

This injury seems self-inflicted in the end and caused by our rack-attack configuration.
